### ISSUE TITLE
[bitnami/mastodon] Change to improve init-job stability broke the init-job

### ISSUE THREAD

**Issue by ajguyot** [bitnami/charts#29982]
### Name and Version

bitnami/mastodon 6.2.10

### What architecture are you using?

amd64

### What steps will reproduce the bug?

Attempt a Helm upgrade from v6.2.9 to v6.2.10

### Are you using any custom parameters or values?

I'm using an external secret for S3, perhaps that's the issue? The broken PR seems to show an expectation that the S3 secret is present in the values.yaml. However the error logs wouldn't make me think this is the problem but I'm not sure how better to debug.

### What is the expected behavior?

The upgrade completes successfully.

### What do you see instead?

init-job pod fails with exit code 127.

Logs state only the following:
```
Defaulted container "migrate-and-create-admin" out of: migrate-and-create-admin, mastodon-assets-precompile, copy-assets-dir (init), wait-for-s3 (init)
Error from server (BadRequest): container "migrate-and-create-admin" in pod "mastodon-init-knms2" is waiting to start: PodInitializing
```

### Additional information

PR that triggered the issue is #28553

**Comment by juan131** [comment]
Hi @ajguyot 

> However the error logs wouldn't make me think this is the problem but I'm not sure how better to debug.

Could you share the output of the command below?

```bash
kubectl logs POD_NAME -c wait-for-s3
```

> I'm using an external secret for S3, perhaps that's the issue?  The broken PR seems to show an expectation that the S3 secret is present in the values.yaml

There's no such as dependency. You can use an external secret for S3 as long as you set the `externalS3.existingSecret` and, optionally, the `externalS3.existingSecretAccessKeyIDKey` and `externalS3.existingSecretKeySecretKey` ones, see:

- https://github.com/bitnami/charts/tree/main/bitnami/mastodon#external-s3-parameters

**Comment by ajguyot** [comment]
Relevant output from a  `kubectl describe` on the init pod, if that helps:
```
Controlled By:  Job/mastodon-init
Init Containers:
[...]
  wait-for-s3:
    Container ID:    containerd://5b4fbd79dd336cc3affc2e3672c14e7bc3c706de6595e7b1d03a185d365e5855
    Image:           docker.io/bitnami/mastodon:4.2.10-debian-12-r0
    Image ID:        docker.io/bitnami/mastodon@sha256:e6ae355b774cadebce1565f5e79cca430f86a77b133755ccde83dea0f8968601
    Port:            <none>
    Host Port:       <none>
    SeccompProfile:  RuntimeDefault
    Command:
      /bin/bash
      -ec
    Args:
      /scripts/wait-for-s3.sh
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    127
      Started:      Wed, 23 Oct 2024 11:15:34 -0500
      Finished:     Wed, 23 Oct 2024 11:15:34 -0500
    Ready:          False
    Restart Count:  6
    Requests:
      cpu:     1
      memory:  1Gi
    Environment:
      BITNAMI_DEBUG:                   false
      MASTODON_S3_HOSTNAME:            s3.us-west-000.backblazeb2.com
      MASTODON_S3_PORT_NUMBER:         443
      MASTODON_AWS_ACCESS_KEY_ID:      <set to the key 'access-key-id' in secret 's3-secret'>      Optional: false
      MASTODON_AWS_SECRET_ACCESS_KEY:  <set to the key 'access-key-secret' in secret 's3-secret'>  Optional: false
    Mounts:
      /scripts from scripts (rw)
      /tmp from empty-dir (rw,path="tmp-dir")
[...]
Volumes:
  empty-dir:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  scripts:
    Type:        ConfigMap (a volume populated by a ConfigMap)
    Name:        mastodon-init-scripts
    Optional:    false
[...]
Events:
  Type     Reason     Age                  From               Message
  ----     ------     ----                 ----               -------
  Normal   Scheduled  10m                  default-scheduler  Successfully assigned mastodon/mastodon-init-hdvft to [node_id]
  Normal   Pulled     10m                  kubelet            Container image "docker.io/bitnami/mastodon:4.2.10-debian-12-r0" already present on machine
  Normal   Created    10m                  kubelet            Created container copy-assets-dir
  Normal   Started    10m                  kubelet            Started container copy-assets-dir
  Normal   Started    9m45s (x4 over 10m)  kubelet            Started container wait-for-s3
  Normal   Pulled     8m51s (x5 over 10m)  kubelet            Container image "docker.io/bitnami/mastodon:4.2.10-debian-12-r0" already present on machine
  Normal   Created    8m51s (x5 over 10m)  kubelet            Created container wait-for-s3
  Warning  BackOff    24s (x48 over 10m)   kubelet            Back-off restarting failed container wait-for-s3 in pod mastodon-init-hdvft_mastodon
```

**Comment by juan131** [comment]
Hi @ajguyot 

Could you please provide the chart parameters (provided via `values.yaml` or using the `--set` flag) you used to install your Helm release? Please exclusively provide the parameters you customized avoiding the ones with default values.

> Note: you can easily obtain the above parameters using `helm get values RELEASE_NAME`

**Comment by joelparker** [comment]
I ran into something similar going from bitnami/mastodon version 8.0.0 to 9.0.1.  My issue was wait-for-backend in the mastodon init job was waiting for ElasticSearch but I had disabled ElasticSearch with

.Values.initJob.migrateElasticsearch = false

In Mastodon version 9.0.0 this moved to:
.Values.initJob.migrateAndCreateAdmin.migrateElasticsearch = false

So the init job was always failing.  

```
kubectl -n test-mastodon logs mastodon-init-c67sw -c wait-for-backend
mastodon 23:04:39.31 INFO  ==> Waiting for PostgreSQL to be ready at test-postgresql:5432/bitnami_mastodon
mastodon 23:04:39.49 INFO  ==> PostgreSQL instance is ready
mastodon 23:04:39.50 INFO  ==> Waiting for Elasticsearch to be ready at http://elasticsearch:9200
```

I fixed my values.yaml and got a new init-job failure:

```
kubectl -n test-mastodon logs mastodon-init-pcgrl
Defaulted container "migrate-and-create-admin" out of: migrate-and-create-admin, mastodon-assets-precompile, copy-assets-dir (init), wait-for-backend (init)
mastodon 00:13:50.46 INFO  ==> Migrating database
mastodon 00:13:50.46 INFO  ==> Waiting for PostgreSQL to be ready at test-postgresql:5432/bitnami_mastodon
mastodon 00:13:50.52 INFO  ==> PostgreSQL instance is ready
mastodon 00:13:54.51 INFO  ==> Creating admin user
Failure/Error: email
    taken
Failure/Error: account.username
    taken
mastodon 00:14:01.83 ERROR ==> Error creating admin user
```

This makes sense on an upgrade (I already have a admin account) and in my old values.yaml had .Values.initJob.createAdmin = false .  Except mastodon_ensure_admin_user_exists will always be written to init-job-configmap.yaml because .Values.initJob.migrateAndCreateAdmin isn't a boolean and should probably be  .Values.initJob.migrateAndCreateAdmin.createAdmin

https://github.com/bitnami/charts/blob/356941ddb5c6ebbd08867aae2b07dfa21ff0e394/bitnami/mastodon/templates/init-job/init-job-configmap.yaml#L50

The documentation talks about .Values.initJob.migrateAndCreateAdmin.createAdmin which is a boolean value:
https://github.com/bitnami/charts/blob/3f157edcff98a7f7c694b043a6ff3e36e0e515b5/bitnami/mastodon/values.yaml#L1431

So initjob always fails for me on an upgrade but works on a clean install when the admin account is created

**Comment by juan131** [comment]
@joelparker thanks so much for the analysis and submitting a PR to address the bug, I just reviewed and merged your contribution.

**Comment by ajguyot** [comment]
Okay I identified my issue, but still haven't solved. The `mastodon-init-scripts` ConfigMap is not getting updated in the Helm upgrade process, so my upgraded init job pod is still using the previous ConfigMap which did not have the `wait-for-s3.sh` script.

I've been researching, but haven't been able to figure out how to get Helm to update this ConfigMap to the new version that should be getting output by version `6.2.10` of this chart. Most of what I've found seems to recommend adding a checksum to make the update take place (https://helm.sh/docs/howto/charts_tips_and_tricks/#automatically-roll-deployments), but if that were what's missing then I'd expect others using this chart to also have had failed upgrades.

Maybe this is a problem on my Helm/K8s setup, in which case I'm sorry for wasting y'all's time here. Will close this out if I figure it out but just wondering if anyone has run into this and knows a solution.

**Comment by ajguyot** [comment]
Okay I was able to manually patch the ConfigMap and then the upgrade went through. Had to do the same for the next version as the ConfigMap changed once again upgrading to 6.2.11.

Still no idea why this is an issue, but I ran `helm template` and the ConfigMap that gets output is indeed correct, so I don't think the problem is in my values.yaml. It may well be an "on my machine" sort of thing (although I'm still wondering if that checksum annotation I linked above should be added to the init job yaml or something along those lines) so I will close this issue out. Thanks for the help!
